README.md
# video_dimensions

Quick and easy video attributes -- width, height, bitrate, codec, duration.
Inspired by Sam Stephenson's [Dimensions](https://github.com/sstephenson/dimensions) gem.

## Features

Seamlessly supports multiple "backends" for sussing out video attributes, so it
will try to find whatever utility the system has available (see
[Supported Backends](#supported-backends)) and use that.

## Examples

```ruby
require 'video_dimensions'

VideoDimensions.dimensions('~/Movies/720p.wmv') # => [1280, 720]
VideoDimensions.width('~/Movies/720p.wmv')      # => 1280
VideoDimensions.height('~/Movies/720p.wmv')     # => 720
VideoDimensions.bitrate('~/Movies/720p.wmv')    # => 5904
VideoDimensions.codec('~/Movies/720p.wmv')      # => "WMV3"
VideoDimensions.duration('~/Movies/720p.wmv')   # => "00:00:02"

v = VideoDimensions('~/Movies/720p.wmv')
v.dimensions # => [1280, 720]
v.width      # => 1280
v.height     # => 720
v.bitrate    # => 5904
v.codec      # => "WMV3"
v.duration   # => "00:00:02"
```

## Requirements

One of the supported backends available in the system's `$PATH`.

### Supported Backends

* [FFmpeg](http://ffmpeg.org/) - `ffmpeg`
* [MediaInfo](http://mediainfo.sourceforge.net/en) - `mediainfo`

## Install

    $ gem install video_dimensions

## Version History

**0.2.0** (2012-08-25)

* Add `duration` attribute

**0.1.1** (2012-08-21)

* Update FFmpeg backend patterns to match more video types

**0.1.0** (2012-08-20)

* Initial release.

## Copyright

Copyright (c) 2012 Robert Speicher

See LICENSE.txt for details.


ChangeLog.md
### 0.1.0 / 2012-08-20

* Initial release:



spec/spec_helper.rb
require 'rspec'
require 'video_dimensions'

Dir[File.expand_path("./spec/support/*.rb")].each { |f| require(f) }

RSpec.configure do |c|
  c.mock_with :mocha

  c.include Fixtures
end


spec/support/string.rb
# https://github.com/mynyml/unindent
class String
  def unindent
    indent = self.split("\n").select {|line| !line.strip.empty? }.map do |line|
      line.index(/[^\s]/)
    end.compact.min || 0

    self.gsub(/^[[:blank:]]{#{indent}}/, '')
  end
end


spec/support/fixtures.rb
module Fixtures
  def fixture(path)
    File.expand_path(File.join('..', 'fixtures', path), File.dirname(__FILE__))
  end
end


lib/video_dimensions.rb
require 'video_dimensions/backends'

# Returns a VideoDimensions instance for determining attributes for the
# provided video.
#
# input - Full path to video being processed
#
# Examples
#
#   v = VideoDimensions('~/Movies/720p.wmv')
#   v.dimensions # => [1280, 720]
#   v.width      # => 1280
#   v.height     # => 720
#   v.bitrate    # => 5904
#   v.codec      # => "WMV3"
#   v.duration   # => "00:00:02"
#
# Returns a VideoDimensions::Backends object
def VideoDimensions(input)
  VideoDimensions::Backends.first_available.new(input)
end

module VideoDimensions

  # Public: Provides shortcuts to attribute methods
  #
  # Examples
  #
  #   VideoDimensions.dimensions('~/Movies/720p.wmv') # => [1280, 720]
  #   VideoDimensions.width('~/Movies/720p.wmv')      # => 1280
  #   VideoDimensions.height('~/Movies/720p.wmv')     # => 720
  #   VideoDimensions.bitrate('~/Movies/720p.wmv')    # => 5904
  #   VideoDimensions.codec('~/Movies/720p.wmv')      # => "WMV3"
  #   VideoDimensions.duration('~/Movies/720p.wmv')   # => "00:00:02"
  class << self
    def dimensions(input)
      process(:dimensions, input)
    end

    def width(input)
      process(:width, input)
    end

    def height(input)
      process(:height, input)
    end

    def bitrate(input)
      process(:bitrate, input)
    end

    def codec(input)
      process(:codec, input)
    end

    def duration(input)
      process(:duration, input)
    end

    private

    def process(msg, input)
      Backends.first_available.new(input).send(msg)
    end
  end

end


lib/video_dimensions/backends/base.rb
module VideoDimensions
  module Backends
    # Public: Base class for Backends
    class Base
      def self.available?
        false
      end

      # Public: Video dimension
      #
      # Returns video dimensions as an array of width and height
      #   in pixels.
      def dimensions
        raise NotImplementedError
      end

      # Public: Video width
      #
      # Returns video width in pixels
      def width
        raise NotImplementedError
      end

      # Public: Video height
      #
      # Returns video height in pixels
      def height
        raise NotImplementedError
      end

      # Public: Video bitrate
      #
      # Returns video bitrate in kbps
      def bitrate
        raise NotImplementedError
      end

      # Public: Video codec
      #
      # Returns video codec ID
      def codec
        raise NotImplementedError
      end

      # Public: Video duration
      #
      # Returns video duration as a string in hh:mm:ss format
      def duration
        raise NotImplementedError
      end
    end
  end
end


lib/video_dimensions/backends/ffmpeg.rb
module VideoDimensions
  module Backends
    class FFmpeg < Base
      def self.available?
        `which #{binary} 2>&1 >> /dev/null`
        $?.exitstatus == 0
      end

      def self.binary
        'ffmpeg'
      end

      def initialize(input)
        @input = input
      end

      # Public: Video dimension
      #
      # Examples
      #
      #   >> FFmpeg.new('720p.wmv').dimensions
      #   => [1280, 720]
      #
      # Returns video dimensions as an array of width and height in pixels.
      def dimensions
        output.match(/Stream .+ Video: .+ (\d+)x(\d+).*/) do |m|
          [m[1].to_i, m[2].to_i]
        end
      end

      # Public: Video width
      #
      # Returns video width in pixels
      def width
        dimensions && dimensions[0]
      end

      # Public: Video height
      #
      # Returns video height in pixels
      def height
        dimensions && dimensions[1]
      end

      # Public: Video bitrate
      #
      # Returns video bitrate in kbps
      def bitrate
        # Video streams don't always surface their own bitrate, so we'll settle
        # for the "total" bitrate
        output.match(/^\s+Duration: .+bitrate: (\d+) kb\/s$/m) do |m|
          m[1].to_i
        end
      end

      # Public: Video codec
      #
      # Examples
      #
      #   >> FFmpeg.new('720p.wmv').codec
      #   => "wmv3"
      #
      # Returns video codec ID
      def codec
        output.match(/Stream .+ Video: (\w+)/m) do |m|
          m[1]
        end
      end

      # Public: Video duration
      #
      # Returns video duration as a string in hh:mm:ss format
      def duration
        output.match(/Duration: ([\d\:]+)/) do |m|
          m[1]
        end
      end

      private

      def output
        unless @output
          @output = `#{self.class.binary} -i "#{@input}" 2>&1`.strip

          # Strip out the Audio codec section(s)
          @output.gsub!(/.*(Input #0.*)output file must be specified.*/m, '\1')
        end

        @output
      end
    end
  end
end


lib/video_dimensions/backends/media_info.rb
module VideoDimensions
  module Backends
    class MediaInfo < Base
      def self.available?
        `which #{binary} 2>&1 >> /dev/null`
        $?.exitstatus == 0
      end

      def self.binary
        'mediainfo'
      end

      def initialize(input)
        @input = input
      end

      # Public: Video dimension
      #
      # Examples
      #
      #   >> MediaInfo.new('720p.wmv').dimensions
      #   => [1280, 720]
      #
      # Returns video dimensions as an array of width and height
      #   in pixels.
      def dimensions
        [width, height] if width && height
      end

      # Public: Video width
      #
      # Returns video width in pixels
      def width
        output.match(/^Width\s+: ([\d\s]+) pixels$/) do |m|
          m[1].gsub(' ', '').to_i
        end
      end

      # Public: Video height
      #
      # Returns video height in pixels
      def height
        output.match(/^Height\s+: ([\d\s]+) pixels$/) do |m|
          m[1].gsub(' ', '').to_i
        end
      end

      # Public: Video bitrate
      #
      # Returns video bitrate in kbps
      def bitrate
        output.match(/^Bit rate\s+: ([\d\s]+) Kbps$/) do |m|
          m[1].gsub(' ', '').to_i
        end
      end

      # Public: Video codec
      #
      # Examples
      #
      #   >> MediaInfo.new('720p.wmv').codec
      #   => "WMV3"
      #
      # Returns video codec ID
      def codec
        output.match(/^Codec ID\s+: (.+)$/) do |m|
          m[1]
        end
      end

      # Public: Video duration
      #
      # Returns video duration as a string in hh:mm:ss format
      def duration
        # MediaInfo only incudes the two longest durations in the ouptut (e.g.,
        # 1h 5m; 5m 10s; 10s 15ms)
        output.match(/^Duration\s+:( \d+h)?( \d+mn)?( \d+s)?(?: \d+ms)?$/) do |m|
          # Map all of the captures to Integers, removing the time notation and
          # converting nils to 0, and use sprintf to make include leading zeros
          sprintf("%02d:%02d:%02d", *m.captures.map(&:to_i))
        end
      end

      private

      def output
        unless @output
          @output = `#{self.class.binary} "#{@input}"`

          # Strip out the Audio codec section(s)
          @output.gsub!(/(.*)^Audio( #\d+)?$.*/m, '\1')
        end

        @output
      end
    end
  end
end
